### PR TITLE
Implementation of App.beep() for Android

### PR DESCRIPTION
**Pull request by t-arn**
This PR implements App.beep() for Android
The "window" example has been extended to demo this feature.

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Comment by t-arn** [comment]
@freakboy3742 Please review this PR

**Comment by mhsmith** [comment]
FYI: rather than waiting for GitHub to run the pre-commit checks, you can set it up on your own machine as shown [here](https://toga.readthedocs.io/en/stable/how-to/contribute-code.html).

File: examples/window/window/app.py [hunk 1]
```diff
@@ -183,9 +189,17 @@ def startup(self):
...
                 btn_change_content,
                 btn_hide,
+                btn_beep,
             ],
             style=Pack(direction=COLUMN),
         )
+        self.scroller = toga.ScrollContainer(
+            horizontal=False,
+            vertical=True,
+            style=Pack(flex=1, padding=10),
```
Comment (freakboy3742) [inline examples/window/window/app.py:199]: This puts the padding on the *outside* of the scroll container, so there's a border around the scrolling content. We don't need any padding there.

File: examples/window/window/app.py [hunk 2]
```diff
@@ -183,9 +189,17 @@ def startup(self):
...
+        )
+        self.scroller.content = self.inner_box
+        self.main_box.add(self.scroller)
```
Comment (freakboy3742) [inline examples/window/window/app.py:202]: The widget layout you've got here is: content -> main_box -> scroller -> inner_box -> *buttons

AFAICT, the main box isn't adding anything. You should be able to add the scroller directly as the main content.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
When I run this on the Briefcase default emulator, I get:
```
D/Ringtone: Successfully created local player
E/FileSource: Failed to open file 'content://settings/system/notification_sound'. (No such file or directory)
E/AudioManager: hasHapticChannels failure:java.io.IOException: Failed to instantiate extractor.
```

This looks like it might be a limitation of the emulator - can you confirm whether you're seeing this as well?

From poking around looking for a quick fix for this, I also saw some suggestions that you should use MediaPlayer, rather than RingtoneManager to play sounds like this, because MediaPlayer will play over any existing sounds (e.g., music playing in the background), rather than interrupting audio.

**Comment by t-arn** [comment]
> E/FileSource: Failed to open file 'content://settings/system/notification_sound'. (No such file or directory)

This is strange. It works on my emulator and I had the code running on my phone as well (in pyPlayground).
I'll try to run the example on the phone again.

**Comment by freakboy3742** [comment]
FWIW: It does beep on my physical Android device; it's only my emulator where it raises an error. Even there, the app doesn't *crash* - it's just a bit noisy in the logs. If we can establish that this is a quirk of (some) emulators, I can live with that; if we can catch the error and log something a little more helpful, that's even better.

File: examples/window/window/app.py (hunk 1 above)
Comment (t-arn) [reply examples/window/window/app.py:199]: ok, removed padding.

File: examples/window/window/app.py (hunk 2 above)
Comment (t-arn) [reply examples/window/window/app.py:202]: ok, removed main_box and renamed scroller to main_scroller, so it becomes obvious that the scroller represents the main content.

**Comment by t-arn** [comment]
> From poking around looking for a quick fix for this, I also saw some suggestions that you should use MediaPlayer, rather than RingtoneManager to play sounds like this, because MediaPlayer will play over any existing sounds (e.g., music playing in the background), rather than interrupting audio.

I think that interrupting existing sounds is what we want in this case: the user **should** get notified. If we play the beep over the running sound, it might go unnoticed by the user.

**Comment by t-arn** [comment]
@freakboy3742 Please re-review this PR.

**Comment by mhsmith** [comment]
I have a few devices of various ages, so I can test this one.

File: android/src/toga_android/app.py [hunk 3]
```diff
@@ -207,7 +210,11 @@ def show_about_dialog(self):
         self.interface.factory.not_implemented("App.show_about_dialog()")
 
     def beep(self):
-        self.interface.factory.not_implemented("App.beep()")
+        uri_obj = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION)
+        uri = java.cast(Uri, uri_obj)
+        ring_obj = RingtoneManager.getRingtone(self.native.getApplicationContext(), uri)
+        ring = java.cast(Ringtone, ring_obj)
+        ring.play()
```
Comment (mhsmith) [inline android/src/toga_android/app.py:217]: I don't think these `cast` calls are necessary. Why did you add them?

**Review by mhsmith — COMMENTED** [review]
I got the same "Failed to open file" error on the Briefcase default emulator with API level 31, and a Pixel 7 with API level 33. But in both cases, the sound still worked, so I think the message is harmless. It's not throwing an exception to us, so there's no way to suppress it.

It also works on an older device and emulator. However, I have noticed that emulators sometimes stop producing any sound at all. You can usually get it back by rebooting the emulator:
* Close the emulator.
* Open Android Studio and go to Tools > Device Manager.
* Open the ellipsis menu next to the emulator in question, and choose "Cold boot now".

The interaction with background music seems reasonable: the notification is clearly audible, but it doesn't interrupt the music any more than necessary.

So I'm fine with this, except for one thing:

File: android/src/toga_android/app.py (hunk 3 above)
Comment (t-arn) [reply android/src/toga_android/app.py:217]: You are right: we already get the correct class instances. Somehow, I thought we get Objects back... I'll fix this.

**Comment by t-arn** [comment]
> I got the same "Failed to open file" error on the Briefcase default emulator with API level 31, and a Pixel 7 with API level 33. But in both cases, the sound still worked, so I think the message is harmless. It's not throwing an exception to us, so there's no way to suppress it.

Do you also have the error message on real devices?

**Comment by mhsmith** [comment]
I got the error message on a real Pixel 7 with API level 33, but not on a Nexus 5X with API level 27. The sound worked on both devices.

**Comment by t-arn** [comment]
> I got the error message on a real Pixel 7 with API level 33, but not on a Nexus 5X with API level 27. The sound worked on both devices.

Yes, the problem with the error message seems to depend on the Android version. With API level 30, there is no error message. Starting with API level 31, the error message appears

**Comment by t-arn** [comment]
@mhsmith Please re-review.

**Comment by t-arn** [comment]
I managed to get rid of the error message on Android 13 :-)

**Review by freakboy3742 — APPROVED** [review]
Looks good; after restarting the emulator as suggested by @mhsmith, I'm now getting a beep (or a twinkle, or whatever you want to call that sound :-). 

Interestingly, I'm *not* getting any file errors any more - I get a warning about offset/length being adjusted:
```
W/FileSource: offset/length adjusted from 0/576460752303423487 to 0/16905
```
but the audio plays fine, so I'll call that a win.

I've renamed the changefile to a `misc` - since the `beep()` API was added in this release, it will already have a feature indicator (#2018); we don't need a separate feature for Android.
